tv_apps/smart-system/js/app_install_manager.js

```diff
@@ -292,6 +325,66 @@ var AppInstallManager = {
     app.onprogress = this.handleProgress;
   },
 
+  previewApp: function ai_previewApp(app) {
+    if (!this.isMarketPlaceAppActive()) {
+      return;
+    }
+
+    var marketplace = AppWindowManager.getActiveApp();
+    var manifestURL = app.manifestURL;
+    var appURL = app.origin + app.manifest.launch_path;
+
+    marketplace.childWindowFactory.createPreviewWindow({
+      detail: {
+        url: appURL,
+        origin: app.origin,
+        manifestURL: manifestURL
+      }
+    });
+
+    marketplace.element.addEventListener('_closed', (evt) => {
+      this.uninstallAllPreviewApps();
+    });
+
+    var previewAppWindow = marketplace.getTopMostWindow();
+    var previewContent = previewAppWindow.element;
+    var msg = {
+      title: 'Press OPTION to',
+      text: 'Add to Apps'
+    };
+
+    previewContent.addEventListener('_opened', (evt) => {
+      var TYPE = window.InteractiveNotifications.TYPE;
+      window.interactiveNotifications.showNotification(TYPE.TOAST, msg);
+    });
+  },
+
+  getAllPreviewAppManifestURL: function() {
+    var previewApps = localStorage.getItem('preview-app') || '';
+    return JSON.parse(previewApps);
+  },
+
+  addPreviewAppManifestURL: function(manifestURL) {
+    var previewApps = this.getAllPreviewAppManifestURL();
+    previewApps[manifestURL] = true;
+    localStorage.setItem('preview-app', JSON.stringify(previewApps));
+  },
+
+  removePreviewAppManifestURL: function(manifestURL) {
+    var previewApps = this.getAllPreviewAppManifestURL();
+    delete previewApps[manifestURL];
```

Comment:
Deleting an element in an array won't release its position. You'll get a `null` in the serialized string.